crypto/x509/v3_purp.c

```diff
@@ -758,51 +771,59 @@ static int no_check(const X509_PURPOSE *xp, const X509 *x, int ca)
 }
 
 /*-
- * Various checks to see if one certificate issued the second.
- * This can be used to prune a set of possible issuer certificates
- * which have been looked up using some simple method such as by
- * subject name.
+ * Various checks to see if one certificate potentially issued the second.
+ * This can be used to prune a set of possible issuer certificates which
+ * have been looked up using some simple method such as by subject name.
  * These are:
  * 1. Check issuer_name(subject) == subject_name(issuer)
  * 2. If akid(subject) exists, check that it matches issuer
  * 3. Check that issuer public key algorithm matches subject signature algorithm
- * 4. If key_usage(issuer) exists, check that it supports certificate signing
```

Comment:
I am not sure we can change API contract in this way. Would there be a different option how to implement this change without breaking API contract for existing 3rd party callers?

Reply:
The original semantics was too strong and the name of the function misleading.
What we can do for backward compatibility is to
* leave the name and semantics of the (original) function as confused as it was and just clarify its documentation,
* split its implementation into 
`int X509_check_potentially issued(X509 *issuer, X509 *subject)` (or use some other name better expressing its semantics) which may be non-public and contains the new semantics according to the current state of the PR (i.e., consisting of the checks 1, 2., and 3 as listed in the code comment shown above)
and the already provided new function
`int X509_signing_allowed(const X509 *issuer, const X509 *subject)` which consists of check 4 only,
* let the new implementation of 
`int X509_check_issued(X509 *issuer, X509 *subject)` be the conjunction of these two, but not use it ourselves, and
* replace the calls of `X509_check_issued()` in the current state of the PR by `X509_check_potentially issued()`, which actually adds to the clarity of the new code.

Reply:
This looks like a good plan. I'd keep the new calls internal.

Reply:
I've nearly finished implementing this part and most other parts of the improvements so far discussed for this PR. I plan to do some more tests and push the resulting major update to this PR tomorrow.

Reply:
Done